optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_xl.py

```diff
@@ -0,0 +1,888 @@
+                        )
+
+                    # compute the previous noisy sample x_t -> x_t-1
+                    latents_batch = self.scheduler.step(
+                        noise_pred, timestep, latents_batch, **extra_step_kwargs, return_dict=False
+                    )[0]
+
+                    if not self.use_hpu_graphs:
+                        self.htcore.mark_step()
+
+                    if callback_on_step_end is not None:
+                        callback_kwargs = {}
+                        for k in callback_on_step_end_tensor_inputs:
+                            callback_kwargs[k] = locals()[k]
+                        callback_outputs = callback_on_step_end(self, i, timestep, callback_kwargs)
+
+                        latents_batch = callback_outputs.pop("latents_batch", latents_batch)
+                        text_embeddings_batch = callback_outputs.pop("text_embeddings_batch", text_embeddings_batch)
+                        add_text_embeddings_batch = callback_outputs.pop(
+                            "add_text_embeddings_batch", add_text_embeddings_batch
+                        )
+                        add_time_ids_batch = callback_outputs.pop("add_time_ids_batch", add_text_embeddings_batch)
+
+                    # call the callback, if provided
+                    if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                        if callback is not None and i % callback_steps == 0:
+                            step_idx = i // getattr(self.scheduler, "order", 1)
+                            callback(step_idx, timestep, latents)
+
+                if not output_type == "latent":
+                    # Post-processing
+                    image = self.vae.decode(latents_batch / self.vae.config.scaling_factor, return_dict=False)[0]
+                else:
+                    image = latents_batch
+
+                outputs["images"].append(image)
+
+                if not self.use_hpu_graphs:
+                    self.htcore.mark_step()
+
+            speed_metrics_prefix = "generation"
+            speed_measures = speed_metrics(
+                split=speed_metrics_prefix,
+                start_time=t0,
+                num_samples=num_batches * batch_size if t1 == t0 else (num_batches - 2) * batch_size,
+                num_steps=num_batches,
+                start_time_after_warmup=t1,
+            )
+            logger.info(f"Speed metrics: {speed_measures}")
+
+            # Remove dummy generations if needed
+            if num_dummy_samples > 0:
+                outputs["images"][-1] = outputs["images"][-1][:-num_dummy_samples]
+
+            # Process generated images
+            for i, image in enumerate(outputs["images"][:]):
+                if i == 0:
+                    outputs["images"].clear()
+
+                image = self.image_processor.postprocess(image, output_type=output_type)
```

Comment:
We should apply the watermark before postprocessing: https://github.com/huggingface/diffusers/blob/v0.23.1/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L1103-L1106

Reply:
Added watermark